### ISSUE TITLE
Support immutable vcpkg root and standard unix filesystem structure

### ISSUE THREAD

**Issue by gracicot** [microsoft/vcpkg#22790]
****

**Is your feature request related to a problem? Please describe.**

Right now vcpkg assumes that `/opt` or `/usr/bin` or wherever vcpkg is installed to be mutable. This is not a problem when you install it manually in `/opt/vcpkg` and set the permissions to your user, but this is a problem when packaging vcpkg.

This is especially a problem when trying to create a nix package for vcpkg, since the nix store is immutable. It's also difficult to install files in a standard unix structure with directories like `bin/` and `share/`

**Proposed solution**

What I would like: be able to install the vcpkg executable under `$prefix/bin/vcpkg` and the data, port, version and everything to `$prefix/share/vcpkg/`. vcpkg should be able to pickup those paths when no `VCPKG_ROOT` is defined, from the path of the executable.

Some of the requirements that would need to support this feature:
 - The vcpkg executable should be able to take filesystem lock under `/var/run/<hash-of-absolute-vcpkg-path>/.vcpkg-root` which is common practice for package manager, and a known mutable location.
 - The vcpkg executable should have a configurable `installed/`, `buildtrees/` and `packages/` locations, or assume a known mutable place like `~/.vcpkg/<hash-of-absolute-vcpkg-path>/` for those directories

Of course, `<hash-of-absolute-vcpkg-path>` can be anything that uniquely identify this particular installation of vcpkg. 

**Describe alternatives you've considered**

The considered alternative for me was to package vcpkg with a wrapper script for the executable in path that takes a hash of the absolute path and create a mutable directory under `~/.vcpkg/<hash-of-absolute-vcpkg-path>/` and simlink everything from inside the `VCPKG_ROOT`, and set the root there instead. I found this alternative brittle. It should be supported natively by vcpkg.

**Additional context**

https://discourse.nixos.org/t/known-mutable-location-in-filesystem-for-lock-files/17337/2

**Comment by gracicot** [comment]
@JackBoosY should I open a discussion in the discussion board?

**Comment by JackBoosY** [comment]
@gracicot Just keep this here.

**Comment by dg0yt** [comment]
> Right now vcpkg assumes that `/opt` or `/usr/bin` or wherever vcpkg is installed to be mutable. This is not a problem when you install it manually in `/opt/vcpkg` and set the permissions to your user, but this is a problem when packaging vcpkg.

This is a particular problems of third-party packaging of vcpkg. As a regular user, you can use vcpkg anywhere in your home directory, as many times as you like.

**Comment by gracicot** [comment]
@dg0yt indeed. This is a problem when packaging vcpkg for distributions. As far as I experienced, ***none*** of the prepackaged vcpkg (which are easier than manually compiling vcpkg) works properly perhaps beside the nix wrapper script I wrote for my own usage. I tried arch's AUR, brew and some other.

Of course, you can go through the manual steps to configure vcpkg, then adding the `VCPKG_ROOT` to your env, then having all build scripts using that environment variable, oh and then you have to update it manually, on all the workstation that you have setup in the past. This is much more tedious than having it pre-packaged, and having automation installing it correctly, and having a package manager keeping it up to date.

Most program can be packaged correctly. Most package managers like npm, pip, cargo, can all be installed using a distribution package, and they will *all* work out of the box. But vcpkg won't. It won't just work out of the box unless you... compile it from source and manually update it?

I don't want to disable users doing the manual setup, or having it as an embedded subproject, I simply want to enable a new way to use it that fits standards and make it easier to automate installation and upgrades.

**Comment by JackBoosY** [comment]
@gracicot Maybe https://github.com/microsoft/vcpkg-ce ?

**Comment by gracicot** [comment]
@JackBoosY I fail to see how it solves vcpkg not being installable and unable to be packaged in other system package managers like brew, nix, pacman or dpkg. They are extremely suited to manage programs and tools on a workstation.

**Comment by dg0yt** [comment]
@gracicot I only commented about your claim that `vcpkg` makes assumptions about certain system paths. It doesn't.
This wasn't a general judgement about the issue but only addressed the quoted part.

> It won't just work out of the box unless you... compile it from source and manually update it?

AFAICT for major platforms it just works out of the box and uses a precompiled binary.
It has got a simple and clear way to update which works on my Ubuntu 18.04 as on my Macbook Air as in MSYS2.
And the Ubuntu 18.04 system packages of other package managers are frozen on old versions.

Note that for versioning in manifest mode, the vcpkg root must be a git checkout, not a plain filesystem.

**Comment by dg0yt** [comment]
> ... other system package managers like brew, nix, pacman or dpkg. They are extremely suited to manage programs and tools on a workstation.

Note that vcpkg doesn't aim to manage programs and tools for users. It is focused on libraries. Pure tools and applications are rejected. (Libraries seem to be accepted regardless of quality or relevance.)
(And I don't say that it has to be this way. It is rather a limitation, because I have to install autotools etc. with a different package manager on each host.)

**Comment by gracicot** [comment]
@dg0yt Oh sorry this isn't what I meant. I meant that system package managers are extremely suited to manage pure tools and applications. Since they are well suited for the task, it's a shame that vcpkg cannot be installed through those. I want to keep vcpkg for C++ libraries, and use AUR, nix, brew to install vcpkg itself.

As you can see from my first comment, it's not possible to install vcpkg on nix without the massive hack of making symlinks of everything in the home directory just so vcpkg agrees to put the downloads there. A simple environment variable to tell vcpkg where it can mutate the filesystem freely would fit x the issue and make vcpkg mostly packagable.

**Comment by gracicot** [comment]
> Note that for versioning in manifest mode, the vcpkg root must be a git checkout, not a plain filesystem.

I see. vcpkg cannot be used when installed as a program like other apps as long as it assume it is used from source (clone the repository I mean, don't really care if it builds something or not)

I can install npm through brew and other system package managers.

I can install pip through nix if I want.

I always installed cargo as a dependency for building a particular package, and my system package manager installed it correctly, and the trust build script was able to use cargo, and I never even touched cargo or its repository.

But vcpkg? No you can't. We must clone the vcpkg repo to use it from there to have all the features.

This is not a problem when developing an app and enjoy vcpkg managing libraries and versions, but this is a ***huge*** problem for packaging an application that depends on vcpkg to be built. I cannot put vcpkg as a build dependency for my package, since package managers can't install vcpkg, your can only clone it for some maybe-required features.

It means that for any app that need to be packaged, it cannot use vcpkg as a build dependency, thus it cannot use vcpkg to download dependencies.

Unless you tell me that when packaging an app that is supposed to be built using vcpkg I should use the dependencies installed by my system package manager, but this is extremely limiting.

**Comment by gracicot** [comment]
> AFAICT for major platforms it just works out of the box and uses a precompiled binary.

Well, doesn't matter that you have to clone the repo instead of using a tool such as pacman, apt, brew, or even chocolatey.

> Note that for versioning in manifest mode, the vcpkg root must be a git checkout, not a plain filesystem.

This is very bad news, but let's focus on the fact that vcpkg assume the filesystem is mutable first.

**Comment by gracicot** [comment]
> AFAICS you can always clone a particular state of the repo, bootstrap, and installed defined versions of the desired dependencies in manifest mode

It's the problem I'm trying to highlight. I don't have to do that with any other package managers out there. If I'm writing an MAKEPKG file, I cannot put `vcpkg` as build dependency in that script and use the manifest to install dependencies. It will fail right now.

I don't have to clone the repository of `pip` to use it, and I don't have to clone the repository of `npm` to use it either. They are installed by my system package manager and they will work out of the box, and I had no problem compiling a package that had `cargo` as builddep. There are no package doing that with vcpkg because it doesn't work with vcpkg, since you have to clone the repo somewhere, then specify where is vcpkg to the build script.

**Comment by gracicot** [comment]
I am writing a useless content to "add new content", sorry for all the notifications I'm causing. This issue is not solved yet.

**Comment by yzx9** [comment]
Once again, adding new content to prevent the issue from being closed.

**Comment by gracicot** [comment]
We should keep this issue open but vcpkg is almost there. The only thing blocking vcpkg to be installed in a immutable door system or of the box is that vcpkg still since it can write a lockfile in the vcpkg root path. All other issues have been mitigated or has a workaround.

**Comment by sketch34** [comment]
+1 I also need a solution for using vcpkg with nix.

**Comment by gracicot** [comment]
@sketch34 Nix have a vcpkg package, just use it. I maintain it and applied a minimal set of patches to make it work in a immutable root.

**Comment by sketch34** [comment]
Thank you @gracicot! I'm brand new to Nix, going through the tutorials now. Looking forward to having a reproducible environment!

**Comment by sketch34** [comment]
@gracicot is there an example using the nix vcpkg package?

**Comment by gracicot** [comment]
Yes. On a machine with nix installed, type the following command:
```
nix shell nixpkgs#vcpkg
```
Use it as you would normally use vcpkg.

You'll now have a shell with vcpkg so you can try it. For using it in project you'll have to write a flake.nix or shell.nix. There's plenty of examples online on how to get started with flakes or nix packages. In the list of nativeBuildInputs, ensure you have vcpkg in there.

**Comment by sketch34** [comment]
> Use it as you would normally use vcpkg.

I normally use vcpkg in manifest mode with a cmake toolchain file. The vcpkg toolchain file is located by CMakeLists.txt using the VCPKG_ROOT env var. 

```
set(CMAKE_TOOLCHAIN_FILE $ENV{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake CACHE STRING "Vcpkg toolchain file")
```

How would I set VCPKG_ROOT for the Nix vcpkg package?

**Comment by gracicot** [comment]
@sketch34 It's up to you, there's many ways to do it. You can follow what I do here: https://github.com/gracicot/kangaru/blob/fb21d3e43ec3783572f0e8c526f66ae8a15ef15c/flake.nix

No there's no need to run it and you won't find that file.

**Comment by sketch34** [comment]
Looks like it might not work with vcpkg manifest mode, even with sandbox disabled.

```
nix-build --keep-failed -A ae --impure --option sandbox false
```

```-- Configuring incomplete, errors occurred!
note: keeping build directory '/tmp/nix-build-ae-v1.0.drv-23'
error: builder for '/nix/store/qxg4vcay8g7nd008k1y9k0bz4hf8480r-ae-v1.0.drv' failed with exit code 1;
       last 25 log lines:
       >   CMAKE_INSTALL_LIBDIR="lib"
       >   CMAKE_INSTALL_PREFIX:PATH="/tmp/nix-build-ae-v1.0.drv-23/engines/build/no-opt/install"
       >
       > -- Engines root: /tmp/nix-build-ae-v1.0.drv-23/engines
       > -- VCPKG_ROOT: /nix/store/49jrzklqhm1d7cbwmcjz9xr4644b6hx3-vcpkg-2024.10.21/share/vcpkg
       > -- VCPKG_INSTALLED_DIR: /tmp/nix-build-ae-v1.0.drv-23/engines/vcpkg_installed
       > -- VCPKG_MANIFEST_DIR: /tmp/nix-build-ae-v1.0.drv-23/engines
       > -- CMAKE_RUNTIME_OUTPUT_DIRECTORY: /tmp/nix-build-ae-v1.0.drv-23/engines/build/no-opt/64bit/bin
       > -- CMAKE_LIBRARY_OUTPUT_DIRECTORY: /tmp/nix-build-ae-v1.0.drv-23/engines/build/no-opt/64bit/lib
       > -- CMAKE_ARCHIVE_OUTPUT_DIRECTORY: /tmp/nix-build-ae-v1.0.drv-23/engines/build/no-opt/64bit/lib
       > -- Running vcpkg install
       > /tmp/vcpkg/e469f7c234a525d45af42b439f85f8d3aada44be1924675428239d8e345887de: error: Failed to take the filesystem lock
       >     Permission denied
       > -- Running vcpkg install - failed```

**Comment by gracicot** [comment]
@sketch34 please report a bug in the nixpkgs repo. I'd like to keep the discussion here about vcpkg itself and UNIX like structure / immutable file system

**Comment by gracicot** [comment]
As an update to this issue: To make vcpkg installable on a read only file system in nix, we apply two patches:
 1. One patch to change the lock file location to be outside of the immutable file system: https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/vc/vcpkg-tool/change-lock-location.patch
 2. A second patch to read the information about the bundle from the root dir instead of the executable dir. This avoid us having to create a directory structure with the bundle and a symlink to the executable. This is primarily because we package vcpkg and vcpkg-tool separately: https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/vc/vcpkg-tool/read-bundle-info-from-root.patch

All other parameters like where to place build trees and installed packages is handled by our wrapper script: https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/vc/vcpkg-tool/package.nix#L109-L171. It's not a perfect wrapping script since it doesn't act perfectly for manifest mode, but it still quite usable.
